lgc/patch/PatchBufferOp.cpp

```diff
@@ -1308,28 +1304,12 @@ void PatchBufferOp::postVisitMemSetInst(MemSetInst &memSetInst) {
 //
 // @param value : The pointer operand value to get as an instruction.
 Value *PatchBufferOp::getPointerOperandAsInst(Value *const value) {
-  // If the value is already an instruction, return it.
-  if (Instruction *const inst = dyn_cast<Instruction>(value))
-    return inst;
-
-  // If the value is a constant (i.e., null pointer), return it.
+  // If the value is a constant (i.e., null pointer), remap it.
   if (isa<Constant>(value)) {
     Constant *const nullPointer = ConstantPointerNull::get(getRemappedType(value->getType()));
     m_replacementMap[value] = std::make_pair(nullptr, nullPointer);
-    return value;
   }
-
-  ConstantExpr *const constExpr = cast<ConstantExpr>(value);
-
-  Instruction *const newInst = m_builder->Insert(constExpr->getAsInstruction());
```

Comment:
Since this function no longer converts constants to instructions, the name "getPointerOperandAsInst" seems wrong. How about getRemappedValue (by analogy with getRemappedType)?

Reply:
The whole thing is a bit strange. Since you're apparently looking at this pass anyway, how about:

* Change this into `getRemappedValue` as suggested and make it return the (descriptor, offset) pair from the map (including an assertion that the source value is already there). The special case for null pointer constants can then be handled without adding an entry to the `m_replacementMap`
* It's weird that `m_replacementMap` serves double duty as replacement map and list of instructions to be erased. This is *especially* confusing when it comes to `load`s (which could hypothetically produce a pointer value!). How about:
  * Add an m_replacedInstructions vector to which all replaced instructions are pushed (including old GEPs).
  * Walk this vector in reverse at the end of `runImpl`, erasing instructions as we go. Add an assertion that erased instructions have no uses anymore. (I think the only potential case not handled are debug instructions?)

Reply:
> * Change this into `getRemappedValue` as suggested and make it return the (descriptor, offset) pair from the map (including an assertion that the source value is already there). The special case for null pointer constants can then be handled without adding an entry to the `m_replacementMap`

Agreed, that works out well. Done.

> * It's weird that `m_replacementMap` serves double duty as replacement map and list of instructions to be erased. This is _especially_ confusing when it comes to `load`s (which could hypothetically produce a pointer value!). How about:
>   
>   * Add an m_replacedInstructions vector to which all replaced instructions are pushed (including old GEPs).
>   * Walk this vector in reverse at the end of `runImpl`, erasing instructions as we go. Add an assertion that erased instructions have no uses anymore. (I think the only potential case not handled are debug instructions?)

I'm less convinced about this part. I'm not sure walking the vector in reverse would work in the presence of phis. Anyway it should be a separate patch.

Reply:
As far as I can tell the code doesn't allow for forward references so wouldn't work correctly in the presence of phis anyway, and it's implicitly relying on some of the restrictions on pointers in SPIR-V. But it's fine not to change this now.

Reply:
Never mind, it does handle incomplete phis. I somehow missed that the first time around.

Reply:
Yes it has a separate late pass for incomplete phis. I don't think this is optimal, e.g. the way it propagates isDivergent and isInvariant info through phis seems like it should use a proper dataflow-style iteration to a fixed point, rather than just two passes. But that is for another patch.